### PR TITLE
RK-312 : Implementing backlink and caching

### PR DESCRIPTION
**Pull request by shinealex33**
RK-312 : Implementing backlink and caching

### CODE REVIEW

**Comment by ModisAtHMRC** [comment]
Within the function `isCacheNotPresentCreateOne` in `SessionCache.scala`, we have the following block processing a `Future[Boolean]:
```
        saveFraudReportDetails(FraudReportDetails(None)) flatMap { 
          case true => fraudReportDetails
          // $COVERAGE-OFF$Loggers
          case false => throw new IllegalStateException("Not able to create cache in Mongo")
          // $COVERAGE-ON
        }
```
The function `saveFraudReportDetails` has this implementation:
```
  def saveFraudReportDetails(rdh: FraudReportDetails)(implicit hc: HeaderCarrier): Future[Boolean] =
    createOrUpdate(sessionId, fraudReportDetailsKey, Json.toJson(rdh)) map (_ => true)
```
It always maps to true, so it appears the case false match is redundant, unless we need to change the implementation to reflect the success status of the operation.

The function it uses, `createOrUpdate`, returns a `Future` containing a `DataBaseUpdate[A],` which appears to log the outcome of the operation within the property `writeResult: LastError`, which in turn has a field `ok: Boolean` that I assume tells us whether the DB op succeeded or not.

I am searching for docs to confirm this but believe this is the value we need to be extracting withing the `map`.

**Comment by ModisAtHMRC** [comment]
In `SessionCache.scala`, the function
```
  def isCachePresent(sessionId: String): Future[Boolean] = findById(sessionId).map {
    case Some(Cache(_, Some(data), _, _)) => true;
    case _                                => false;
  }
```
can be shortened to
```
  def isCachePresent(sessionId: String): Future[Boolean] = findById(sessionId) map {
    _ exists { cache => cache.data.isDefined }
  }
```.

**Comment by ModisAtHMRC** [comment]
In `ActivityTypeController.scala`, the method `onPageLoad` can be made a little more readable from
```scala
 Action.async {
    implicit request =>
      sessionCache.isCacheNotPresentCreateOne(hc.sessionId.get.value) flatMap {
        case fraudReport if fraudReport.activityType.isDefined =>
          Future.successful(
            Ok(activityTypeView(form.fill(ActivityTypeViewModel(fraudReport.activityType.get)), onSubmitActivityType()))
          )
        case _ => Future.successful(Ok(activityTypeView(form, onSubmitActivityType())))

      }
  }
```
to
```scala
 Action.async {
    implicit request =>
      Future.successful {
        Ok {
          sessionCache.isCacheNotPresentCreateOne(hc.sessionId.get.value) flatMap { fraudReport =>
            val filledForm = fraudReport.activityType map {
              form.fill( ActivityTypeViewModel( _ )
            } getOrElse form

            activityTypeView( filledForm, onSubmitActivityType() )
          }
        }
      }
  }
```

**Comment by ModisAtHMRC** [comment]
`UserAnswersCacheSpec.scala` appears to have redundant imports of `TimeUnit`.

**Comment by ModisAtHMRC** [comment]
`ActivityTypeControllerSpec.scala` appears to have a redundant `SessionId` important.

**Comment by shinealex33** [comment]
> In `SessionCache.scala`, the function
> 
> ```scala
>   def isCachePresent(sessionId: String): Future[Boolean] = findById(sessionId).map {
>     case Some(Cache(_, Some(data), _, _)) => true;
>     case _                                => false;
>   }
> ```
> 
> can be shortened to
> 
> ```scala
>   def isCachePresent(sessionId: String): Future[Boolean] = findById(sessionId) map {
>     _ exists { cache => cache.data.isDefined }
>   }
> ```

**Comment by shinealex33** [comment]
> 

resolved. Please take 

> Within the function `isCacheNotPresentCreateOne` in `SessionCache.scala`, we have the following block processing a `Future[Boolean]`:
> 
> ```scala
>         saveFraudReportDetails(FraudReportDetails(None)) flatMap { 
>           case true => fraudReportDetails
>           // $COVERAGE-OFF$Loggers
>           case false => throw new IllegalStateException("Not able to create cache in Mongo")
>           // $COVERAGE-ON
>         }
> ```
> 
> The function `saveFraudReportDetails` has this implementation:
> 
> ```scala
>   def saveFraudReportDetails(rdh: FraudReportDetails)(implicit hc: HeaderCarrier): Future[Boolean] =
>     createOrUpdate(sessionId, fraudReportDetailsKey, Json.toJson(rdh)) map (_ => true)
> ```
> 
> It always maps to `true`, so the `case false` match appears redundant, unless we need to change the implementation to reflect the success status of the operation.
> 
> The function it uses, `createOrUpdate`, returns a `Future` containing a `DataBaseUpdate[A],` which appears to log the outcome of the operation within the property `writeResult: LastError`, which in turn has a field `ok: Boolean` that I assume tells us whether the DB op succeeded or not.
> 
> I am searching for docs to confirm this but believe this is the value we need to be extracting within the `map`.



> Within the function `isCacheNotPresentCreateOne` in `SessionCache.scala`, we have the following block processing a `Future[Boolean]`:
> 
> ```scala
>         saveFraudReportDetails(FraudReportDetails(None)) flatMap { 
>           case true => fraudReportDetails
>           // $COVERAGE-OFF$Loggers
>           case false => throw new IllegalStateException("Not able to create cache in Mongo")
>           // $COVERAGE-ON
>         }
> ```
> 
> The function `saveFraudReportDetails` has this implementation:
> 
> ```scala
>   def saveFraudReportDetails(rdh: FraudReportDetails)(implicit hc: HeaderCarrier): Future[Boolean] =
>     createOrUpdate(sessionId, fraudReportDetailsKey, Json.toJson(rdh)) map (_ => true)
> ```
> 
> It always maps to `true`, so the `case false` match appears redundant, unless we need to change the implementation to reflect the success status of the operation.
> 
> The function it uses, `createOrUpdate`, returns a `Future` containing a `DataBaseUpdate[A],` which appears to log the outcome of the operation within the property `writeResult: LastError`, which in turn has a field `ok: Boolean` that I assume tells us whether the DB op succeeded or not.
> 
> I am searching for docs to confirm this but believe this is the value we need to be extracting within the `map`.

Fixed the issue - please fix it.

**Comment by shinealex33** [comment]
> 

Fixed issue - please take a look

**Comment by shinealex33** [comment]
> > 
> 
> resolved. Please take
> 
> > Within the function `isCacheNotPresentCreateOne` in `SessionCache.scala`, we have the following block processing a `Future[Boolean]`:
> > ```scala
> >         saveFraudReportDetails(FraudReportDetails(None)) flatMap { 
> >           case true => fraudReportDetails
> >           // $COVERAGE-OFF$Loggers
> >           case false => throw new IllegalStateException("Not able to create cache in Mongo")
> >           // $COVERAGE-ON
> >         }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The function `saveFraudReportDetails` has this implementation:
> > ```scala
> >   def saveFraudReportDetails(rdh: FraudReportDetails)(implicit hc: HeaderCarrier): Future[Boolean] =
> >     createOrUpdate(sessionId, fraudReportDetailsKey, Json.toJson(rdh)) map (_ => true)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It always maps to `true`, so the `case false` match appears redundant, unless we need to change the implementation to reflect the success status of the operation.
> > The function it uses, `createOrUpdate`, returns a `Future` containing a `DataBaseUpdate[A],` which appears to log the outcome of the operation within the property `writeResult: LastError`, which in turn has a field `ok: Boolean` that I assume tells us whether the DB op succeeded or not.
> > I am searching for docs to confirm this but believe this is the value we need to be extracting within the `map`.
> 
> > Within the function `isCacheNotPresentCreateOne` in `SessionCache.scala`, we have the following block processing a `Future[Boolean]`:
> > ```scala
> >         saveFraudReportDetails(FraudReportDetails(None)) flatMap { 
> >           case true => fraudReportDetails
> >           // $COVERAGE-OFF$Loggers
> >           case false => throw new IllegalStateException("Not able to create cache in Mongo")
> >           // $COVERAGE-ON
> >         }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The function `saveFraudReportDetails` has this implementation:
> > ```scala
> >   def saveFraudReportDetails(rdh: FraudReportDetails)(implicit hc: HeaderCarrier): Future[Boolean] =
> >     createOrUpdate(sessionId, fraudReportDetailsKey, Json.toJson(rdh)) map (_ => true)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It always maps to `true`, so the `case false` match appears redundant, unless we need to change the implementation to reflect the success status of the operation.
> > The function it uses, `createOrUpdate`, returns a `Future` containing a `DataBaseUpdate[A],` which appears to log the outcome of the operation within the property `writeResult: LastError`, which in turn has a field `ok: Boolean` that I assume tells us whether the DB op succeeded or not.
> > I am searching for docs to confirm this but believe this is the value we need to be extracting within the `map`.
> 
> Fixed the issue - please fix it.

Fixed issue - please take a look

**Comment by shinealex33** [comment]
> > > 
> > 
> > 
> > resolved. Please take
> > > Within the function `isCacheNotPresentCreateOne` in `SessionCache.scala`, we have the following block processing a `Future[Boolean]`:
> > > ```scala
> > >         saveFraudReportDetails(FraudReportDetails(None)) flatMap { 
> > >           case true => fraudReportDetails
> > >           // $COVERAGE-OFF$Loggers
> > >           case false => throw new IllegalStateException("Not able to create cache in Mongo")
> > >           // $COVERAGE-ON
> > >         }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > The function `saveFraudReportDetails` has this implementation:
> > > ```scala
> > >   def saveFraudReportDetails(rdh: FraudReportDetails)(implicit hc: HeaderCarrier): Future[Boolean] =
> > >     createOrUpdate(sessionId, fraudReportDetailsKey, Json.toJson(rdh)) map (_ => true)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > It always maps to `true`, so the `case false` match appears redundant, unless we need to change the implementation to reflect the success status of the operation.
> > > The function it uses, `createOrUpdate`, returns a `Future` containing a `DataBaseUpdate[A],` which appears to log the outcome of the operation within the property `writeResult: LastError`, which in turn has a field `ok: Boolean` that I assume tells us whether the DB op succeeded or not.
> > > I am searching for docs to confirm this but believe this is the value we need to be extracting within the `map`.
> > 
> > 
> > > Within the function `isCacheNotPresentCreateOne` in `SessionCache.scala`, we have the following block processing a `Future[Boolean]`:
> > > ```scala
> > >         saveFraudReportDetails(FraudReportDetails(None)) flatMap { 
> > >           case true => fraudReportDetails
> > >           // $COVERAGE-OFF$Loggers
> > >           case false => throw new IllegalStateException("Not able to create cache in Mongo")
> > >           // $COVERAGE-ON
> > >         }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > The function `saveFraudReportDetails` has this implementation:
> > > ```scala
> > >   def saveFraudReportDetails(rdh: FraudReportDetails)(implicit hc: HeaderCarrier): Future[Boolean] =
> > >     createOrUpdate(sessionId, fraudReportDetailsKey, Json.toJson(rdh)) map (_ => true)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > It always maps to `true`, so the `case false` match appears redundant, unless we need to change the implementation to reflect the success status of the operation.
> > > The function it uses, `createOrUpdate`, returns a `Future` containing a `DataBaseUpdate[A],` which appears to log the outcome of the operation within the property `writeResult: LastError`, which in turn has a field `ok: Boolean` that I assume tells us whether the DB op succeeded or not.
> > > I am searching for docs to confirm this but believe this is the value we need to be extracting within the `map`.
> > 
> > 
> 
> Fixed issue - please take a look

**Comment by shinealex33** [comment]
> > 
> 
> resolved. Please take
> 
> > Within the function `isCacheNotPresentCreateOne` in `SessionCache.scala`, we have the following block processing a `Future[Boolean]`:
> > ```scala
> >         saveFraudReportDetails(FraudReportDetails(None)) flatMap { 
> >           case true => fraudReportDetails
> >           // $COVERAGE-OFF$Loggers
> >           case false => throw new IllegalStateException("Not able to create cache in Mongo")
> >           // $COVERAGE-ON
> >         }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The function `saveFraudReportDetails` has this implementation:
> > ```scala
> >   def saveFraudReportDetails(rdh: FraudReportDetails)(implicit hc: HeaderCarrier): Future[Boolean] =
> >     createOrUpdate(sessionId, fraudReportDetailsKey, Json.toJson(rdh)) map (_ => true)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It always maps to `true`, so the `case false` match appears redundant, unless we need to change the implementation to reflect the success status of the operation.
> > The function it uses, `createOrUpdate`, returns a `Future` containing a `DataBaseUpdate[A],` which appears to log the outcome of the operation within the property `writeResult: LastError`, which in turn has a field `ok: Boolean` that I assume tells us whether the DB op succeeded or not.
> > I am searching for docs to confirm this but believe this is the value we need to be extracting within the `map`.
> 
> > Within the function `isCacheNotPresentCreateOne` in `SessionCache.scala`, we have the following block processing a `Future[Boolean]`:
> > ```scala
> >         saveFraudReportDetails(FraudReportDetails(None)) flatMap { 
> >           case true => fraudReportDetails
> >           // $COVERAGE-OFF$Loggers
> >           case false => throw new IllegalStateException("Not able to create cache in Mongo")
> >           // $COVERAGE-ON
> >         }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The function `saveFraudReportDetails` has this implementation:
> > ```scala
> >   def saveFraudReportDetails(rdh: FraudReportDetails)(implicit hc: HeaderCarrier): Future[Boolean] =
> >     createOrUpdate(sessionId, fraudReportDetailsKey, Json.toJson(rdh)) map (_ => true)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It always maps to `true`, so the `case false` match appears redundant, unless we need to change the implementation to reflect the success status of the operation.
> > The function it uses, `createOrUpdate`, returns a `Future` containing a `DataBaseUpdate[A],` which appears to log the outcome of the operation within the property `writeResult: LastError`, which in turn has a field `ok: Boolean` that I assume tells us whether the DB op succeeded or not.
> > I am searching for docs to confirm this but believe this is the value we need to be extracting within the `map`.
> 
> Fixed the issue - please fix it.

File: app/uk/gov/hmrc/taxfraudreportingfrontend/controllers/ActivityTypeController.scala [hunk 1]
```diff
@@ -17,40 +17,53 @@
...
-            Redirect(uk.gov.hmrc.taxfraudreportingfrontend.controllers.routes.ReportingTypeController.onPageLoad())
+        activityType =>
+          userAnswersCache.cacheActivityType(activityType.activityType) flatMap (
+            _ =>
+              Future.successful(
+                Redirect(uk.gov.hmrc.taxfraudreportingfrontend.controllers.routes.ReportingTypeController.onPageLoad())
+              )
           )
       )
   }
```
Comment (ModisAtHMRC) [inline app/uk/gov/hmrc/taxfraudreportingfrontend/controllers/ActivityTypeController.scala:69]: The `flatMap` of a `Future.succecssful` can be simplified to a `map`:
```scala
  def onSubmit(): Action[AnyContent] = Action.async {
   import uk.gov.hmrc.taxfraudreportingfrontend.controllers.routes

    implicit request =>
      form.bindFromRequest().fold(
        formWithErrors => Future.successful(
          BadRequest( activityTypeView( formWithErrors, onSubmitActivityType() ) )
        ),
        activityType =>
          userAnswersCache.cacheActivityType(activityType.activityType) map (
            _ => Redirect( routes.ReportingTypeController.onPageLoad() )
          )
      )
  }
```

File: app/uk/gov/hmrc/taxfraudreportingfrontend/controllers/IndexViewController.scala [hunk 2]
```diff
@@ -24,15 +24,19 @@ import uk.gov.hmrc.taxfraudreportingfrontend.config.AppConfig
...
+  implicit val ec: ExecutionContext
 ) extends FrontendController(mcc) {
 
   def onPageLoad(): Action[AnyContent] = Action.async { implicit request =>
-    Future.successful(Ok(indexView()).addingToSession(SessionKeys.sessionId -> UUID.randomUUID.toString))
+    if (request.session.data.contains(SessionKeys.sessionId))
+      Future.successful(Ok(indexView()))
+    else
+      Future.successful(Ok(indexView()).addingToSession(SessionKeys.sessionId -> UUID.randomUUID.toString))
   }
```
Comment (ModisAtHMRC) [inline app/uk/gov/hmrc/taxfraudreportingfrontend/controllers/IndexViewController.scala:40]: The async function can be simplified to a synchronous one:
```scala
  def onPageLoad(): Action[AnyContent] = Action { implicit request =>
    val response = Ok( indexView() )
    
    if ( request.session.data contains SessionKeys.sessionId ) response
    else response.addingToSession(SessionKeys.sessionId -> UUID.randomUUID.toString)
  }
```

**Review by ModisAtHMRC — CHANGES_REQUESTED** [review]
Overall good code which can just be shortened a little by removing some redundant instances of `Future.successful`.

File: app/uk/gov/hmrc/taxfraudreportingfrontend/controllers/ActivityTypeController.scala (hunk 1 above)
Comment (shinealex33) [reply app/uk/gov/hmrc/taxfraudreportingfrontend/controllers/ActivityTypeController.scala:66]: implemented thanks!

File: app/uk/gov/hmrc/taxfraudreportingfrontend/controllers/IndexViewController.scala (hunk 2 above)
Comment (shinealex33) [reply app/uk/gov/hmrc/taxfraudreportingfrontend/controllers/IndexViewController.scala:40]: implemented thanks!

**Review by ModisAtHMRC — APPROVED** [review]
Awesome, futures all sorted. Happy to approve.

**Pull request merged**